flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateReaderImpl.java

```diff
@@ -0,0 +1,116 @@
+ *     <li>buffers must be prepared (cleared) before passing to reader</li>
+ *     <li>buffers must be released after use</li>
+ * </ol>
+ */
+@NotThreadSafe
+@Internal
+public class ChannelStateReaderImpl implements ChannelStateReader {
+	private static final Logger log = LoggerFactory.getLogger(ChannelStateReaderImpl.class);
+
+	private final Map<InputChannelInfo, ChannelStateStreamReader> inputChannelHandleReaders;
+	private final Map<ResultSubpartitionInfo, ChannelStateStreamReader> resultSubpartitionHandleReaders;
+
+	public ChannelStateReaderImpl(TaskStateSnapshot snapshot) {
+		this(snapshot, new ChannelStateSerializerImpl());
+	}
+
+	@SuppressWarnings("WeakerAccess")
+	ChannelStateReaderImpl(TaskStateSnapshot snapshot, ChannelStateDeserializer serializer) {
+		RefCountingFSDataInputStreamFactory streamFactory = new RefCountingFSDataInputStreamFactory(serializer);
+		final HashMap<InputChannelInfo, ChannelStateStreamReader> inputChannelHandleReadersTmp = new HashMap<>();
+		final HashMap<ResultSubpartitionInfo, ChannelStateStreamReader> resultSubpartitionHandleReadersTmp = new HashMap<>();
+		for (Map.Entry<OperatorID, OperatorSubtaskState> e : snapshot.getSubtaskStateMappings()) {
+			addReaders(inputChannelHandleReadersTmp, e.getValue().getInputChannelState(), streamFactory);
+			addReaders(resultSubpartitionHandleReadersTmp, e.getValue().getResultSubpartitionState(), streamFactory);
+		}
+		inputChannelHandleReaders = inputChannelHandleReadersTmp;
+		resultSubpartitionHandleReaders = resultSubpartitionHandleReadersTmp;
+	}
+
+	private <T> void addReaders(Map<T, ChannelStateStreamReader> readerMap, Collection<? extends AbstractChannelStateHandle<T>> handles, RefCountingFSDataInputStreamFactory streamFactory) {
+		for (AbstractChannelStateHandle<T> handle : handles) {
+			checkState(!readerMap.containsKey(handle.getInfo()), "multiple states exist for channel: " + handle.getInfo());
+			readerMap.put(handle.getInfo(), new ChannelStateStreamReader(handle, streamFactory));
+		}
+	}
+
+	@Override
+	public ReadResult readInputData(InputChannelInfo info, Buffer buffer) throws IOException {
+		log.debug("readInputData, resultSubpartitionInfo: {} , bufferBuilder {}", info, buffer);
+		return getReader(info, inputChannelHandleReaders).readInto(buffer);
+	}
+
+	@Override
+	public ReadResult readOutputData(ResultSubpartitionInfo info, BufferBuilder bufferBuilder) throws IOException {
+		log.debug("readOutputData, resultSubpartitionInfo: {} , bufferBuilder {}", info, bufferBuilder);
+		return getReader(info, resultSubpartitionHandleReaders).readInto(bufferBuilder);
+	}
+
+	private <K> ChannelStateStreamReader getReader(K info, Map<K, ChannelStateStreamReader> readerMap) {
+		Preconditions.checkArgument(readerMap.containsKey(info), String.format("unknown channel %s. Known channels: %s", info, readerMap.keySet()));
+		return readerMap.get(info);
+	}
+
+	@Override
+	public void close() throws Exception {
+		try (Closer closer = Closer.create()) {
```

Comment:
we can use the following for simple:
```
IOUtils.closeAll(inputChannelHandleReaders.values());
IOUtils.closeAll(resultSubpartitionHandleReaders.values());
```

Reply:
Each of these two actions can still throw an exception; if both throw then one should be suppressed. This logic to handle that seems more complex to me than nested loops in `try`.

Reply:
Yes, you are right. 
My previous concern was whether it is proper way to rely on `shaded.guava18` for this purpose (I am not quite sure). And we already introduced many basic functions in `IOUtils` for the similar usages. If the current option might also be widely used in other places, maybe it is worth the efforts to also be covered by `IOUtils`.